blogs/ulysses-offload/README.md

```diff
@@ -0,0 +1,263 @@
+Additionally, FPDT is compatible with advanced memory optimization
+techniques such as DeepSpeed ZeRO and PyTorch FSDP, further enhancing
+its usability and performance.
+
+## Core Design of Fully Pipelined Distributed Transformer
+
+Figure 2 illustrates the core structure of FPDT. FPDT leverages multiple
+memory hierarchies in modern GPU clusters, thus boosting hardware
+efficiency and cost-effectiveness while achieving very high model FLOP
+utilization (MFU). The design of FPDT centers around pipelining,
+scheduling, and memory management. These well-known optimization
+techniques are essential for scaling LLM context length to a million
+scale with a few GPUs and will be discussed in the subsequent
+subsections.
+
+<img src="./media/image2.png" style="width:6.5in;height:2.68634in"
+alt="A screenshot of a computer Description automatically generated" />
+
+Figure 2: FPDT core design
+
+### 
+
+### Pipelining and Scheduling
+
+FPDT employs a pipelined sequence chunking design to manage the memory
+and computational load efficiently. In traditional Transformer model,
+input QKV tensor can be denoted *\[B, S, H, D\]*, where *B* is batch
+size, *S* is sequence length, *H* is number of heads and *D* is hidden
+dimension per head. With sequence parallelism such as DeepSpeed Ulysses,
+input tensor is partitioned along sequence dimension across sequence
+parallel group P, that is *\[B, S/P, H,D\]* prior to alltoall collective
+communication. The alltoall communication gathers partitioned tensor
+along sequence dimension and scatter them along head dimension essential
+transforming tensor from *\[B, S/P, H, D\]* to *\[B,S, H/P, D\]*. In our
+FPDT design, we further subdivide per GPU *S/P* sequence into *u*
+chunks. Thus, the input tensor is now represented as \[*B, S/uP, H,
+D*\]. We denote these chunks as *T<sub>i</sub>*,
+where$\ i\  \in \ 0,1,\ldots,\ u - 1.$ As shown in Figure 1,
+*T<sub>i</sub>* is projected to query *q<sub>i</sub>*, key
+*k<sub>i</sub>*, and value *v<sub>i</sub>*. Then, we perform the
+alltoall communication among the sequence parallel group. In our chunk
+design, the sequence length for each chunk is reduced by a factor of *u*
+compared to Ulysses.
+
+<img src="./media/image3.png" style="width:6.5in;height:5.36042in"
+alt="A screenshot of a computer Description automatically generated" />
+
+Figure 3: FDPT with Offload
+
+Figure 3 gives an example of how to perform the computation of chunk
+*T<sub>m</sub>*. After the alltoall collective communication,
+*GPU<sub>j</sub>* receives
+$\widehat{q}m,\ \widehat{k}m,\ and\ \widehat{v}m$*.* We then fetch the
+previous sequence chunk by chunk from the host memory to
+GPU<sub>j</sub>, and perform online attention with the current
+$\widehat{q}m$ and update the output chunk accordingly. Note that, in a
+strict manner, at any given time, only one set of chunks
+$\widehat{k}i,\ and\ \widehat{v}i\ $is placed on GPU's HBM, reducing the
```

Comment:
This is not rendering correctly
<img width="722" alt="image" src="https://github.com/user-attachments/assets/c18721d4-0ce7-4f89-843c-831e6ed0ac89">